File: gitdot-api/src/endpoint/organization/list_organizations.rs
use serde::{Deserialize, Serialize};

use crate::{
    endpoint::Endpoint,
    resource::{OrganizationResource, common::Page},
};

pub struct ListOrganizations;

impl Endpoint for ListOrganizations {
    const PATH: &'static str = "/organizations";
    const METHOD: http::Method = http::Method::GET;

    type Request = ListOrganizationsRequest;
    type Response = ListOrganizationsResponse;
}

#[derive(ApiRequest, Debug, Default, Serialize, Deserialize)]
pub struct ListOrganizationsRequest {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub cursor: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub limit: Option<u32>,
}

pub type ListOrganizationsResponse = Page<OrganizationResource>;

File: gitdot-web/packages/gitdot-api/src/endpoint/organization/list-organizations.ts
import { z } from "zod";
import { OrganizationResource, page } from "../../resource";
import type { Endpoint } from "../endpoint";

export const ListOrganizationsRequest = z.object({
  cursor: z.string().optional(),
  limit: z.number().int().positive().optional(),
});
export type ListOrganizationsRequest = z.infer<typeof ListOrganizationsRequest>;

export const ListOrganizationsResponse = page(OrganizationResource);
export type ListOrganizationsResponse = z.infer<
  typeof ListOrganizationsResponse
>;

export const ListOrganizations = {
  path: "/organizations",
  method: "GET",
  request: ListOrganizationsRequest,
  response: ListOrganizationsResponse,
} as const satisfies Endpoint;
export type ListOrganizations = typeof ListOrganizations;
